src/zjs_callbacks.c

```diff
@@ -315,17 +312,21 @@ void zjs_remove_callback(int32_t id)
     }
 }
 
-void zjs_signal_callback(int32_t id)
+void zjs_signal_callback(int32_t id, void* args, uint32_t size)
 {
-    if (id != -1 && cb_map[id]) {
-#ifdef DEBUG_BUILD
-        if (cb_map[id]->type == CALLBACK_TYPE_JS) {
-            DBG_PRINT("signaling JS callback id %ld\n", id);
-        } else {
-            DBG_PRINT("signaling C callback id %ld\n", id);
+    if (args && size) {
+        // signal with args
+        int ret = zjs_port_ring_buf_put(&ring_buffer, (uint16_t)id, 0, (uint32_t*)args, (uint8_t)size / 4);
+        if (ret != 0) {
+            DBG_PRINT("error putting into ring buffer, ret=%u\n", ret);
+        }
+    } else {
+        // signal without args
+        uint32_t dummy = 0;
+        int ret = zjs_port_ring_buf_put(&ring_buffer, (uint16_t)id, 0, (uint32_t*)&dummy, (uint8_t)1);
```

Comment:
I don't think you can put in "nothing" into the ring buffer. The Zephyr implementation actually doesn't check for NULL and de-references the pointer, so that will segfault, but I don't think passing in no data was a use case for their ring buffer implementation.